index.js

```diff
@@ -1,93 +1,138 @@
+    if (info) this.info = info
+    if (cause) this.cause = cause
+
+    /** @private @type {Array<TaggedError>} */
+    this._oErrorTags // eslint-disable-line
+  }
+
+  /**
+   * Set the extra info object for this error.
+   *
+   * @param {Object | null | undefined} info extra data to attach to the error
+   * @return {this}
+   */
+  withInfo(info) {
+    this.info = info
+    return this
   }
 
   /**
    * Wrap the given error, which caused this error.
    *
-   * @param {Error} cause
+   * @param {Error} cause the internal error that caused this error
    * @return {this}
    */
   withCause(cause) {
     this.cause = cause
-    if (this.message && cause.message) {
-      this.message += ': ' + cause.message
-    }
     return this
   }
-}
 
-/**
- * Return the `info` property from `error` and recursively merge the `info`
- * properties from the error's causes, if any.
- *
- * If a property is repeated, the first one in the cause chain wins.
- *
- * @param {?Error} error assumed not to have circular causes
- * @return {Object}
- */
-function getFullInfo(error) {
-  if (!error) return {}
-  const info = getFullInfo(error.cause)
-  if (typeof error.info === 'object') Object.assign(info, error.info)
-  return info
-}
+  /**
+   * Tag debugging information onto any error (whether an OError or not) and
+   * return it.
+   *
+   * @param {Error} error the error to tag
+   * @param {string} [message] message with which to tag `error`
+   * @param {Object} [info] extra data with wich to tag `error`
+   * @return {Error} the modified `error` argument
+   */
+  static tag(error, message, info) {
+    const oError = /** @type{OError} */ (error)
 
-/**
- * Return the `stack` property from `error` and recursively append the `stack`
- * properties from the error's causes, if any.
- *
- * @param {?Error} error assumed not to have circular causes
- * @return {string}
- */
-function getFullStack(error) {
-  if (!error) return ''
-  const causeStack = getFullStack(error.cause)
-  if (causeStack) return error.stack + '\ncaused by: ' + causeStack
-  return error.stack
+    if (!oError._oErrorTags) oError._oErrorTags = []
+
+    let tag
+    if (Error.captureStackTrace) {
+      // Hide this function in the stack trace, and avoid capturing it twice.
+      tag = /** @type TaggedError */ ({ name: 'TaggedError', message, info })
+      Error.captureStackTrace(tag, OError.tag)
+    } else {
+      tag = new TaggedError(message, info)
+    }
+
+    oError._oErrorTags.push(tag)
+
+    return error
+  }
+
+  /**
+   * The merged info from any `tag`s on the given error.
+   *
+   * If an info property is repeated, the last one wins.
+   *
+   * @param {Error | null | undefined} error any errror (may or may not be an `OError`)
+   * @return {Object}
+   */
+  static getFullInfo(error) {
+    const info = {}
+
+    if (!error) return info
+
+    const oError = /** @type{OError} */ (error)
+
+    if (typeof oError.info === 'object') Object.assign(info, oError.info)
```

Comment:
just noticed that you have this check here but not below. it is not really needed because Object.assign silently ignores invalid inputs, but wondering if there was a reason to have it here but not below.

Reply:
I tried removing it, but it turns out that `Object.assign` will happily assign things over a String, with strange results, so it does add some safety.

The rationale for checking here and not below is that `oError` here might be any kind of error, including someone else's error with an `info` property that isn't what we're expecting. When looping through `_oErrorTags`, we know (or at least can more confidently assume) that the objects there are ours and have the `info` properties we're expecting.

I guess there's an argument for instead calling the property `_oErrorInfo`, analogous to `_oErrorTags`.